### PR TITLE
Add Standardize input transform

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR adds the possibility to standardize inputs for a GP via implementing a `Standardize` input transform. As the `Normalize` input transform it is possible to transform only specific input columns via providing a list of indices. This makes it possible to combine it with other transforms or intentionally ignore inputs in the transform.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/pytorch/botorch, and link to your PR here.)

### CODE REVIEW

**Review by Balandat — COMMENTED** [review]
This looks great, thanks for the PR! Thanks also for the extensive tests. 

Mostly nits, and one comment  on a potential flaky test issue.

File: botorch/models/transforms/input.py [hunk 1]
```diff
@@ -458,6 +458,136 @@ def equals(self, other: InputTransform) -> bool:
         return False
 
 
+class Standardize(ReversibleInputTransform, Module):
```
Comment (saitcakmak) [inline botorch/models/transforms/input.py:461]: How about using a slightly different class name here, just so we can distinguish the two if both the input and output transforms `Standardize` are imported? You can always rename one of them at the import but it may be easier to avoid it altogether. Maybe something like `InputStandardize`?

Also, just curious: Any particular use cases where you would prefer this over `Normalize`?

**Review by qingfeng10 — CHANGES_REQUESTED** [review]
Thanks for setting up the PR! I left some comments. Overall it looks great.

File: botorch/models/transforms/input.py [hunk 2]
```diff
@@ -458,6 +458,136 @@ def equals(self, other: InputTransform) -> bool:
...
+        if self.training:
+            if X.size(-1) != self.means.size(-1):
+                raise BotorchTensorDimensionError(
+                    f"Wrong input. dimension. Received {X.size(-1)}, "
+                    f"expected {self.means.size(-1)}"
+                )
+            self.means = X.mean(dim=-2, keepdim=True)
+            self.stds = X.std(dim=-2, keepdim=True)
+
+            self.stds[torch.where(self.stds <= self.min_std)] = self.min_std
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py:545]: Yep, looks better. We should maybe then do the same for normalize.

File: test/models/transforms/test_input.py [hunk 3]
```diff
@@ -303,6 +304,135 @@ def test_normalize(self):
...
+            self.assertEqual(stdz.stds.shape, torch.Size([1, 2]))
+            self.assertEqual(len(stdz.indices), 1)
+            self.assertTrue((stdz.indices == torch.tensor([0], dtype=torch.long)).all())
+            stdz = Standardize(d=2, indices=[0], batch_shape=torch.Size([3]))
+            self.assertTrue(stdz.training)
+            self.assertEqual(stdz._d, 2)
+            self.assertEqual(stdz.means.shape, torch.Size([3, 1, 2]))
+            self.assertEqual(stdz.stds.shape, torch.Size([3, 1, 2]))
+            self.assertEqual(len(stdz.indices), 1)
+            self.assertTrue((stdz.indices == torch.tensor([0], dtype=torch.long)).all())
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply test/models/transforms/test_input.py]: Fixed it.

File: test/models/transforms/test_input.py [hunk 4]
```diff
@@ -303,6 +304,135 @@ def test_normalize(self):
...
+                self.assertTrue(torch.all(X_stdz.mean(dim=-2).abs() < 1e-4))
+                self.assertTrue(torch.all(X_stdz.std(dim=-2) < 1.0 + 1e-4))
+                self.assertTrue(torch.all(X_stdz.std(dim=-2) > 1.0 - 1e-4))
+                stdz.eval()
+                X_unstdz = stdz.untransform(X_stdz)
+                self.assertTrue(torch.allclose(X, X_unstdz, atol=1e-4, rtol=1e-4))
+                expected_means = X.mean(dim=-2, keepdim=True)
+                expected_stds = X.std(dim=-2, keepdim=True)
+                self.assertTrue(torch.allclose(stdz.means, expected_means))
+                self.assertTrue(torch.allclose(stdz.stds, expected_stds))
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply test/models/transforms/test_input.py:371]: Thanks for spotting. I fixed the seed of the random generator, but I am not sure it was the best way how to fix it. Maybe you can have a look ;)

File: botorch/models/transforms/input.py (hunk 1 above)
Comment (jduerholt) [reply botorch/models/transforms/input.py]: Hi, 

I changed the name to InputStandardize. Thanks for suggesting it!

I had in the past a use case in which using a standard scaler leads to slightly improved predictive performance. But this was more an outlier. Normally, I also use `Normalize`. As I am in the moment in the process of pushing my custom botorch addons upstream, I thought it could be worth of having at least the opportunity. 

Of course, there is no ultimative need for these kind of scalers, as you can scale your `X` also in advance and untransform the proposed `X_prop` after the optimization. But I really like the possibility of uniting scaler and model as this makes it easier to work for example with linear constraints (no need to transform them). 

Best,

Johannes

[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py]: Yes, you are totally correct about this. I changed it!

File: botorch/models/transforms/input.py [hunk 5]
```diff
@@ -458,6 +458,136 @@ def equals(self, other: InputTransform) -> bool:
...
+            raise ValueError("`indices` list is empty!")
+        if (indices is not None) and (len(indices) > 0):
+            indices = torch.tensor(indices, dtype=torch.long)
+            if len(indices) > d:
+                raise ValueError("Can provide at most `d` indices!")
+            if (indices > d - 1).any():
+                raise ValueError("Elements of `indices` have to be smaller than `d`!")
+            if len(indices.unique()) != len(indices):
+                raise ValueError("Elements of `indices` tensor must be unique!")
+            self.indices = indices
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py:509]: I also thought about it because from a purely mathematical point of view `Standardize` and `Normalize` are almost the same, both are linear transformations of the form: `X_transformed = (X-subtrahend)/divisor`.

Only difference is that its coefficients (`subtrahend ` and `divisor`) are calculated differently. A possible solution would be introducing an additional class called `LinearTransform` and inherit then from this. This could look like this:

```python
class LinearTransform(ReversibleInputTransform, Module):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        coefficients: Optional(Tuple(Tensor,Tensor)) = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_divisor: float = 1e-8,
    ) -> None:
        super().__init__()
        if (indices is not None) and (len(indices) == 0):
            raise ValueError("`indices` list is empty!")
        if (indices is not None) and (len(indices) > 0):
            indices = torch.tensor(indices, dtype=torch.long)
            if len(indices) > d:
                raise ValueError("Can provide at most `d` indices!")
            if (indices > d - 1).any():
                raise ValueError("Elements of `indices` have to be smaller than `d`!")
            if len(indices.unique()) != len(indices):
                raise ValueError("Elements of `indices` tensor must be unique!")
            self.indices = indices
        if coefficients is not None:
            if coefficients[0].size(-1) != d:
                raise BotorchTensorDimensionError(
                    "Dimensions of provided subtrahend are incompatible with `d`!"
                )

            if coefficients[1].size(-1) != d:
                raise BotorchTensorDimensionError(
                    "Dimensions of provided divisor are incompatible with `d`!"
                )
            subtrahend = self.coefficients[0]
            divisor = self.coefficients[1]
            self.learn_coefficients = False
        else:
            subtrahend = torch.zeros(*batch_shape, 1, d)
            divisor = torch.zeros(*batch_shape, 1, d)
            self.learn_bounds = True
        self.register_buffer("subtrahend", subtrahend)
        self.register_buffer("divisor", divisor)
        self._d = d
        self.transform_on_train = transform_on_train
        self.transform_on_eval = transform_on_eval
        self.transform_on_fantasize = transform_on_fantasize
        self.reverse = reverse
        self.batch_shape = batch_shape
        self.min_divisor = min_divisor

    def _transform(self, X: Tensor) -> Tensor:
        if self.learn_coefficients and self.training:
            if X.size(-1) != self.divisor.size(-1):
                raise BotorchTensorDimensionError(
                    f"Wrong input dimension. Received {X.size(-1)}, "
                    f"expected {self.divisor.size(-1)}."
                )
            self.subtrahend, self.divisor = self._learn_coefficients()
            self.divisor[torch.where(self.divisor <= self.min_divisor)] = self.min_divisor
        
        if hasattr(self, "indices"):
            X_new = X.clone()
            X_new[..., self.indices] = (
                X_new[..., self.indices] - self.subtrahend[..., self.indices]
            ) / self.divisor[..., self.indices]
            return X_new
        return (X - self.subtrahend) / self.divisor

    def _untransform(self, X: Tensor) -> Tensor:
        if hasattr(self, "indices"):
            X_new = X.clone()
            X_new[..., self.indices] = (
                self.subtrahend[..., self.indices]
                + X_new[..., self.indices] * self.divisor[..., self.indices]
            )
            return X_new
        return self.subtrahend + X * self.divisor

    @abstractmethod
    def learn_coefficients(self):
        return subtrahend, divisor


class Normalize(LinearTransform):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        bounds: Optional[Tensor] = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_range: float = 1e-8,
    ) -> None:
        if bounds is not None:
            subtrahend = bounds[..., 0:1, :]
            divisor = bounds[..., 1:2, :] - subtrahend
        super().__init__(
            d=d,
            indices=indices,
            coefficients=(subtrahend,divisor),
            batch_shape=batch_shape,
            transform_on_train=transform_on_train,
            transform_on_eval=transform_on_eval,
            transform_on_fantasize=transform_on_fantasize,
            reverse=reverse,
            min_divisor=min_range)
        return

    def learn_coefficients(self):
        mins = X.min(dim=-2, keepdim=True)[0]
        ranges = X.max(dim=-2, keepdim=True)[0] - mins
        return mins, ranges


class InputStandardize(LinearTransform):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_std: float = 1e-8,
    ) -> None:
        super().__init__(
            d=d,
            indices=indices,
            coefficients=None,
            batch_shape=batch_shape,
            transform_on_train=transform_on_train,
            transform_on_eval=transform_on_eval,
            transform_on_fantasize=transform_on_fantasize,
            reverse=reverse,
            min_divisor=min_std)
        return

    def learn_coefficients(self):
        means = X.mean(dim=-2, keepdim=True)
        stds = X.std(dim=-2, keepdim=True)
        return means, stds
```

Note this code is not tested or properly formatted and should only solve as an example! Maybe one could integrate something like this in a separate PR as you suggested.

[the remark this answers is not in the document]
Comment (Balandat) [reply botorch/models/transforms/input.py:509]: yeah it's a nice idea to factor out that functionality and consolidate it into a common base class like this. Let's do this in a separate PR. Would you be willing to submit one?

File: test/models/transforms/test_input.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Balandat) [reply test/models/transforms/test_input.py:371]: Lgtm. Six by nine.
